src/parse/asp/builtins.go

```diff
@@ -753,6 +754,62 @@ func addDep(s *scope, args []pyObject) pyObject {
 	return None
 }
 
+func addDatumToTargetAndMaybeQueue(s *scope, target *core.BuildTarget, datum core.BuildInput, systemAllowed, tool bool) {
+	target.AddDatum(datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+func addNamedDatumToTargetAndMaybeQueue(s *scope, name string, target *core.BuildTarget, datum core.BuildInput, systemAllowed, tool bool) {
+	target.AddNamedDatum(name, datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+// Add runtime dependencies to target
+func addData(s *scope, args []pyObject) pyObject {
+	s.Assert(s.Callback, "can only be called from a pre- or post-build callback")
+
+	label := args[0]
+	target := getTargetPost(s, string(label.(pyString)))
+
+	systemAllowed := false
+	tool := false
+
+	// add_data() builtin can take a string, list, or dict
+	if isType(args[1], "str") {
+		if bi := parseBuildInput(s, args[1], string(label.(pyString)), systemAllowed, tool); bi != nil {
+			addDatumToTargetAndMaybeQueue(s, target, bi, systemAllowed, tool)
+		}
+	} else if isType(args[1], "list") {
+		for _, str := range args[1].(pyList) {
+			if bi := parseBuildInput(s, str, string(label.(pyString)), systemAllowed, tool); bi != nil {
+				addDatumToTargetAndMaybeQueue(s, target, bi, systemAllowed, tool)
+			}
+		}
+	} else if isType(args[1], "dict") {
+		for name, v := range args[1].(pyDict) {
+			for _, str := range v.(pyList) {
+				if bi := parseBuildInput(s, str, string(label.(pyString)), systemAllowed, tool); bi != nil {
+					addNamedDatumToTargetAndMaybeQueue(s, name, target, bi, systemAllowed, tool)
+				}
+			}
+		}
+	} else {
+		log.Fatal("Unrecognised data type passed to add_data")
+	}
+
+	// TODO(peterebden): Do we even need the following any more?
```

Comment:
oof yeah I think I tried to get rid of this a while back... Sort of think we shouldn't but fair TODO :)